src/libopensc/pkcs15-pubkey.c

```diff
@@ -1081,6 +1081,8 @@ sc_pkcs15_dup_pubkey(struct sc_context *ctx, struct sc_pkcs15_pubkey *key, struc
 	pubkey->algorithm = key->algorithm;
 
 	if (key->alg_id) {
+		if (key->algorithm == SC_ALGORITHM_GOSTR3410)
+			key->alg_id->params = &(key->u.gostr3410.params);
```

Comment:
Could you please check whether we could improve initializing of the data structures instead?